File: Talleres_2C_2025/taller1/extra/l-linda.py
def movimientos_para_obtener_aLinda(cadena, long): 
    return movimientos_para_llinda(cadena, long, "a")

def movimientos_para_llinda(cadena, long, l):
    #el caso base cuando el string es de 1 caracter, si es l no hace movimientos, si no, necesita 1 movimiento
    if long == 1 and cadena == l:
            return 0
    if long == 1 and cadena != l:
            return 1
    
    #divido la longitud del string para encarar por el lado de D&C, necesito verificar de a mitades
    medio = long//2
    cadenaI = cadena[:medio]    #lado izquierdo del string
    cadenaD = cadena[medio:]    #lado derecho del string

    siguiente_char = chr(ord(l) + 1) #mi l+1

    #caso donde la primera mitad es todo l
    #cuento cuantos caracteres son distintos a l en la primera mitad (suma un mov cada uno), y recursivamente
    #verifico cuantos  necesito en la otra mitad, con el char siguiente a l
    movimientos1 = cantidad_char_distintos(cadenaI, l) + movimientos_para_llinda(cadenaD, medio, siguiente_char) 
    #caso donde la segunda mitad es todo a
    #hago llo mismo pero contanod que todos los caracteres de la segunda mitad sean l
    movimientos2 = cantidad_char_distintos(cadenaD, l) + movimientos_para_llinda(cadenaI, medio, siguiente_char)

    return min(movimientos1, movimientos2)  #devuelve la opcion de cambios que haya necesitado menos movimientos

def cantidad_char_distintos(cadena, l):
    res = 0

    for i in range(0, len(cadena)):
        if cadena[i] != l:
            res += 1
    return res

def main():
    t = int(input())

    for _ in range(t):
        long = int(input())
        cadena = input().strip()

        movimientos_necesarios = movimientos_para_obtener_aLinda(cadena, long)
        print(movimientos_necesarios)

if __name__ == "__main__":
    main()

File: Talleres_2C_2025/taller1/Ej1/l-linda.cpp
/*
tengo un string s[1...n] que es l-lindo si:
o bien |s| = 1, y ese unico caracter es l
o bien |s| > 1, la primera mitad de s esta compuesta solo por l, y la segunda mitad es un (l + 1)lindo
o bien |s| > 1, la primera mitad de s es un (l + 1)lindo, y la segunda mitad esta compuesta solo por l
la longitud de las cadenas siempre va a ser alguna potencia de 2

lo voy a trbajar con el método de divide&conquer

la idea es, primero ver si el string tiene una sola letra, si esa es "a", no necesita ningun cambio,
pero si es otra necesita 1. Ese seria mi caso base
Despues analizar los dos casos restantes, dividiendo la cadena a la mitad para poder calcular
cuantos cambios necesito hacer en una mitad para que sea todo "a", y cuantos cambios necesita la
otra mitad para cumplir con las caracteristicas de palabra (l+1)-linda (calculado recursivamente).
Luego, al unir los movimientos de las dos mitades, me quedan dos posibles casos de movimientos 
necesarios (el caso donde la primera mitad es "a" o cuando la segunda miad es "a"), pero como me pide
la menor cantidad de cambios necesarios, pido el minimo entre esos dos.
*/

#include <iostream>
#include <string>
using namespace std;

//calcula cuantos char en una cadena son distintos al char pasado por parametro
int cantidad_char_distintos(string cadena, char l) {
    
    int res = 0;

    for (int i = 0; i < cadena.length(); i++) {
        if (cadena[i] != l) {
            res++;
        }
    }
    return res;
}

//calcula los movimientos necesarios para transformar una cadena en l-linda valida, usando D&C
int movimientos_para_obtener_llinda(string cadena, int longitud, char l) {

    //el caso base cuando el string es de 1 caracter, si es l no hace cambios, 
    //si no, necesita 1 movimiento para cambiar ese caracter a l
    if (longitud == 1 && cadena[0] == l) {
        return 0;
   }
    if (longitud == 1 && cadena[0] != l) {
        return 1;
    }

    int medio = longitud / 2;
    string cadenaI = cadena.substr(0, medio);
    string cadenaD = cadena.substr(medio);

    //caso donde la primera mitad es todo l
    int movimientos_caso_1 = cantidad_char_distintos(cadenaI, l) + movimientos_para_obtener_llinda(cadenaD, medio, l + 1);
    //caso donde la primera mitad es todo l
    int movimientos_caso_2 = cantidad_char_distintos(cadenaD, l) + movimientos_para_obtener_llinda(cadenaI, medio, l + 1);

    //devuelve la opcion de cambios que haya necesitado menos movimientos
    return min(movimientos_caso_1, movimientos_caso_2); 
}


int main() {

    int t;
    cin >> t;

    for (int i = 0; i < t; i++) {
        int longitud;
        string cadena;

        cin >> longitud;
        cin >> cadena;

        int movimientos_necesarios = movimientos_para_obtener_llinda(cadena, longitud, 'a');
        cout << movimientos_necesarios << endl;
    }

    return 0;
}